FixHorizon changes

Omastto1 opened the pull request:
Hello,
I made a couple of changes to the code to pass the tests and work with the `fixhorizon` utility properly.
I also tried to complete and improve Docstrings.
@zsunberg 
This finalizes your changes. If it is ok, we can do merge #5 also.

zsunberg committed:
first shot at fixhorizon is done

zsunberg committed:
tests pass

Omastto1 committed:
Fixed typo and duplicity, #4

Omastto1 committed:
removed global variable for stage and renamed it

Omastto1 committed:
Fixed documentation

Omastto1 committed:
Fixed stage system

Originally x stages meant x states and x - 1 actions. This commit changes meaning to x + 1 states and x actions-

Omastto1 committed:
Updated methods to work with FHWwapper

Omastto1 committed:
Changed stage_states to product of states and stage

Omastto1 committed:
Fixed tests to work pass state check

Omastto1 committed:
Fixed indexing

Omastto1 committed:
Updated test problem instance definition and test

Omastto1 committed:
Minor bug fix and renaming

Omastto1 committed:
Updated test

Omastto1 committed:
Merge pull request #1 from Omastto1/fixhorizonfixes

UpdateFHExample -> FixHorizonFixes

Omastto1 committed:
Revert "UpdateFHExample -> FixHorizonFixes"

Omastto1 committed:
Merge pull request #2 from Omastto1/revert-1-fixhorizonfixes

Revert "UpdateFHExample -> FixHorizonFixes"

Omastto1 committed:
Merge pull request #3 from Omastto1/updateFHExample

Update fh example -> fix horizon fixes

Omastto1 committed:
removed return in one liner method

Omastto1 committed:
DocStrings improvement

Omastto1 committed:
Test dependencies added

Omastto1 committed:
Merge branch 'fixhorizonfixes'

Omastto1 committed:
Merge branch 'master' of https://github.com/Omastto1/FiniteHorizonPOMDPs.jl

Omastto1 committed:
Fixed merge errors

Omastto1 committed:
fixed redundant definition

Omastto1 committed:
removed deprecated include

BoZenKhaa commented:
Hello, I tested this version and the current version in master. This version is definitely more useable as it does not require specific fields in the MDP struct.

However, I think this change drops support for a class of MDPs that have transition probabilities dependent on the current epoch. 
The `fixhorizon` wrapper fixes a horizon for a general MDP, but unlike the previous version it wouldn't work for an MDP that already has an epoch as a state variable.

I imagine there could be two possible use cases for this package:

1. Take existing mdp with epoch-indepndent transition function, e.g. simple grid world -> apply `fhmdp=fixhorizon(mdp::MDP, horizon::Int64)` to get mdp type with fixed horizon -> apply solver to the FH MDP: `solve(fhmdp::FHWrapper)`

2. Take custom mdp with epoch being a state variable -> implement the FH interface, namely functions `horizon(mdp::MDP), epoch(s::State)` and possibly `stage_states(mdp::MDP, epoch::Int64)` and `stage_states_index(mdp::MDP, s::State, epoch::Int64)` as well -> following the pattern from dicrete value iteration, initialize FH solver type with `solver = FiniteHorizonSolver(;kwargs...)` and apply solve to the mdp, `solve(solver::FiniteHorizonSolver, mdp::MDP)`

I think both usecases could leverage the same algorithm with `solve` dispatching on different arguments. 

Overall, I think that keeping the option for custom finite horizon MDPs is pretty important and I don't see how that could be easily done with the `fixhorizon` approach.

@Omastto1, @zsunberg, what do you think?

Omastto1 committed:
Updated interface with stage(ss::Tuple{<:Any, Int})

Omastto1 commented:
After a discussion with @BoZenKhaa, we agreed that the second use case is, in other words, user-defined Finite Horizon MDP. And as such, it is not necessary to implement some wrapper or so.

This could be solved by broadening `solve`'s definition to something like `solve(w::Union{FHWrapper, MDP})` with user-defined MDP containing interface methods.

Omastto1 commented:
However, if we allow the user to define their own FH MDPs (which is definitely desirable) we should somehow update the `HorizonLength`, because this will break it.

Omastto1 committed:
Updated DocStrings

Omastto1 commented:
Commit f9c8fe26f9709b230e25fa32354f46d60a710aee adds function `stage` to interface.
It replaces use of `last(ss)` in `fixhorizon.jl`.

zsunberg commented:
Hi, sorry for the late response - looking at this and thinking about it now. I will have more comments later this evening (Mountain time USA)

zsunberg commented on src/fixhorizon.jl:
```diff
@@ -1,4 +1,9 @@
 # TODO: Docstring
+"""
+    fixhorizon(m::FHWrapper, horizon::Int)
```
```suggestion
    fixhorizon(m::Union{MDP,POMDP}, horizon::Int)
```

zsunberg commented on src/interface.jl:
```diff
@@ -1,12 +1,23 @@
 
-# TODO: Docstring
+"""
+    HorizonLength(::Type{<:FHWrapper})
```
```suggestion
    HorizonLength(::Type{<:Union{POMDP,MDP})
    HorizonLength(::Union{POMDP,MDP})
```
`HorizonLength(MyPOMDP) == FiniteHorizon()` means that every object of type MyPOMDP has a finite horizon. `HorizonLength(m) == FiniteHorizon()` means that the specific (PO)MDP object `m` has a finite horizon (this could depend on an object parameter. Does that makes sense to have both? I think it does.

Omastto1 replied:
It definitely does.
I will commit these changes.
On the other hand, this is connected with #15. 
My opinion on this is that we should allow the user to implement his own FH (PO)MDP, but we should also force him either to use something to distinguish FH MDPs from IH MDPs or his definition should be of a specific type.
This way we will be able to differentiate between `HorizonLength(::Type{<:Union{POMDP, MDP})` and user-defined FH MDPs.
I think that this should be the topic of the next version (even if smaller), not this one.

Omastto1 replied:
Going through code I realized that the user should also implement his own `HorizonLength(::Type{<:MyPOMDP})` method. That solves the issue and thus does not have to be dealt with.

zsunberg commented on src/fixhorizon.jl:
```diff
@@ -25,43 +30,120 @@ const FHWrapper = Union{FixedHorizonMDPWrapper,FixedHorizonPOMDPWrapper}
+    HorizonLength(::Type{<:FHWrapper})
+
+Check whether MDP is Finite or Infinite Horizon and return corresponding struct (FiniteHorizon or InfiniteHorizon).
+"""
```
Why do we need this docstring here? It is documented in interface.jl

zsunberg commented on src/fixhorizon.jl:
```diff
@@ -25,43 +30,120 @@ const FHWrapper = Union{FixedHorizonMDPWrapper,FixedHorizonPOMDPWrapper}
+
+Return number of state's stage
+"""
+stage(ss::Tuple{<:Any,Int}) = last(ss)
```
This should be documented in interface.jl - this docstring seems redundant to me. Same for the functions below.

Omastto1 replied:
I was not sure how to write docstrings in cases of declaration in the interface and with the definition in `fixhorizon`.
The usage with input parameters in the docstring in the interface is bad, so I wanted to make explicit docstring in `fixhorizon`.
Anyway, I fixed this.

zsunberg commented on src/solver.jl:
```diff
@@ -9,61 +10,51 @@
     action_map::Vector{A}
     include_Q::Bool 
-    mdp::M
+    w::W
```
I think this should be able to hold any MDP, right? As in #15 a user should be able to define their own finite horizon MDPs and they shouldn't have to use an FHWrapper. Same below.

zsunberg replied:
Of course, the solver should throw an error if `HorizonLength(m) == InfiniteHorizon()`.

zsunberg commented on src/solver.jl:
```diff
@@ -9,61 +10,51 @@
 end
 
+# MDP given horizon 5 assumes that agent can move 5 times
+function solve(w::FHWrapper; verbose::Bool=false, new_VI::Bool=true)
```
As you suggested in #14, I think this should be

```suggestion
function POMDPs.solve(solver::FiniteHorizonSolver, m::MDP)
```

zsunberg commented on src/valueiteration.jl:
```diff
@@ -1,18 +1,21 @@
+function valueiterationsolver(w::FHWrapper, stage::Int64, util)
     next_stage_value = util               # maximum value in each row
-    stage_q = fill(0., (mdp.no_states, length(actions(mdp)))) # q_matrix preinitialization # XXX: use length(stage_states(mdp)) instead of mdp.no_states?
+    stage_q = fill(0., (length(stage_states(w, 1)), length(actions(w))))
```
yes, this is much better!

zsunberg reviewed (changes requested):
I liked a lot of the changes that you made!

I left some comments mostly relating to #14 and #15. Let me know when you've addressed all these and/or if you have a reason to keep it the way it is, and then we can merge soon! @BoZenKhaa good to have you looking over this too!!

Omastto1 committed:
Update src/fixhorizon.jl

Co-authored-by: Zachary Sunberg <zachary.sunberg@colorado.edu>

Omastto1 committed:
Update src/interface.jl

Update `HorizonLength` input arguments in docstring

Co-authored-by: Zachary Sunberg <zachary.sunberg@colorado.edu>

Omastto1 committed:
Fixed docstrings

Omastto1 committed:
Merge branch 'master' of https://github.com/Omastto1/FiniteHorizonPOMDPs.jl

Omastto1 committed:
Added FHSolver and support for user-defined FHMDPs

Omastto1 commented:
I reviewed your proposals and added support for user-defined FH MDPs and added `FiniteHorizonSolver` (I think we should change the name in the future, assuming I will add FH POMDPs solver).

I am yet to properly test the user-defined POMDPs with the package though. I also added a test for non-FH MDP input.

Omastto1 committed:
Fixed custom FHMDPs bug

Omastto1 committed:
Added tests for Custom FH MDPs

Omastto1 commented:
Dug up a former custom instance of FH MDP and tested it with the current code, which works ok.

Omastto1 committed:
Added FH VI solver requirements

Omastto1 committed:
Added methods implementation checker for fixhorizon test

zsunberg commented:
changed this to merge to master instead of fixhorizon.

Cool, do you think we are ready to merge @Omastto1 ? I will try to take a quick look tomorrow and if there is nothing to change, I will merge it.

BoZenKhaa commented on src/interface.jl:
```diff
@@ -1,12 +1,24 @@
-# TODO: Improve docstring https://docs.julialang.org/en/v1/manual/documentation/
-"stage_states(mdp::MDP, epoch::Int64)"
+"""
```
I think it would be great if usage examples were shown in the interface documentation, e.g. 
```
"""
     stage(s::MDPState)::Integer
Return number of state's stage
"""
````

Omastto1 replied:
The documentation contained examples few commits back, then you pointed out that the functions in the interface do not have input arguments specified, so I removed them.

BoZenKhaa replied:
Oh ok, I think that might have been a misunderstanding. I took issue with some of the examples having the very specific input types, such as FHWrapper. My point was that the example should not be so specific if the function is supposed to be defined by the suer for some generic user-specified mdp. Sorry about the confusion, I in fact really like such usage examples.

BoZenKhaa commented on src/solver.jl:
```diff
@@ -9,79 +10,106 @@
 # Policy struct created according to one used in DiscreteValueIteration
+
+#TODO: Dosctring
+struct FiniteHorizonSolver <: Solver
```
It would be beneficial to add constructor function that takes the parameter as a keyword.

BoZenKhaa commented on src/solver.jl:
```diff
@@ -9,79 +10,106 @@
 
-        if verbose
-            println("EPOCH: $epoch")
+    for stage=m.horizon:-1:1
```
`horizon(m)` should be used in this file instead of `m.horizon`.

BoZenKhaa commented on test/custom1dtest.jl:
```diff
@@ -22,13 +27,15 @@ solver = ValueIterationSolver(max_iterations=10, belres=1e-3, include_Q=true);
 
 # Solve Finite Horizon by Value Iteration
-FHPolicy = FiniteHorizonPOMDPs.solve(mdp);
+FHPolicy = FiniteHorizonPOMDPs.solve(fhsolver, mdp);
```
As a usage example, I think it would be better to call `POMDPs.solve` here instead

Omastto1 replied:
If I understand this correctly, you are saying that instead of `FiniteHorizonPOMDPs.solve(fhsolver, mdp)` I should use `POMDPs.solve(fhsolver, mdp)`. Correct?
I am calling specific packages to call the desired methods to compare them.

BoZenKhaa replied:
My point is coming from the point of view of someone who looks to tests to see how to use the package. As you are defining `solve` as a new method of `POMDPs.solve` and you export it, I think it should be fine just to use `solve(fhsolver, mdp)` here. Anyway, my point is that I was confused by seeing `FiniteHorizonPOMDPs` here as I saw elsewhere that you define it as a new method for `POMDPs.solve`. But you can leave this as is, I don't think it's important.

BoZenKhaa commented:
I have tested this on my own FHMDP instance and I am getting a different actions between the FH solver and the VI solver. This wasn't happening with the original version of the code, but it may be an error on my end. Previously I think I was using the `stage_stateindex(mdp::PMDP, s::State)` instead of `stage_stateindex(mdp::PMDP, s::State, epoch::Int64)`, so I think I might have broken something there. 

I don't have time to chase this down now right now, but I will try to figure it out soon or at least create a minimum working example.

BoZenKhaa commented on src/solver.jl:
```diff
@@ -9,79 +10,106 @@
             println("util")
             println(util)
             println("policy")
             println(policy)
```
Should be `fhpolicy`. There probably should be a test where the `verbose` flag is true.

Omastto1 replied:
I agree this needs an update. What is your proposal for such a test? Wouldn't it be more visual test than an automated one?

BoZenKhaa replied:
I think just having a small test where you *run* the code and don't test for anything with `verbose=true` would be enough, just to see that the code is not throwing exceptions (which it does in this case as `policy` is undefined).

Omastto1 replied:
Good point, would not notice that bug without you, thank you!

BoZenKhaa commented on src/solver.jl:
```diff
@@ -9,79 +10,106 @@
 
 function action(policy::FiniteHorizonValuePolicy, s::S) where S
-    sidx = stage_stateindex(policy.mdp, s, s.epoch)
+    sidx = stateindex(policy.m, s)
```
With my MDP, this was causing issues. If I follow this right, `stateindex` is using the "global" mdp indexing while `stage_stateindex` is using a "local" stage indexing. Because the solution algorithm works on the stage indexes, fetching action based on `stateindex` from the "global" indexing will get a wrong action. 

So this line should imo be:
`sidx = stage_stateindex(policy.mdp, s, stage(s))`

Maybe a more complex 2D testacase where there is larger number of actions being used would help catch stuff like this?

BoZenKhaa replied:
Just a note, changing to `sidx = stage_stateindex(policy.mdp, s, stage(s))` fixes my problem in https://github.com/JuliaPOMDP/FiniteHorizonPOMDPs.jl/pull/8#issuecomment-775271381, but I am not sure if it's not causing problems elsewhere.

zsunberg replied:
> Just a note, changing to `sidx = stage_stateindex(policy.mdp, s, stage(s))` fixes my problem in [#8 (comment)](https://github.com/JuliaPOMDP/FiniteHorizonPOMDPs.jl/pull/8#issuecomment-775271381), but I am not sure if it's not causing problems elsewhere.

I strongly think we should NOT have the three argument `stage_stateindex(mdp, s, stage)`. The stage information should be included in the state.

Omastto1 replied:
I think the current `stateindex` is the correct solution. The `policy.policy` stores policy for each state, so you also need a global index of state.
`policy.policy` is `horizon x (no_states + 1)` big. Getting local index would only obtain actions from first row/column.
Please provide example.

Omastto1 replied:
> > Just a note, changing to `sidx = stage_stateindex(policy.mdp, s, stage(s))` fixes my problem in [#8 (comment)](https://github.com/JuliaPOMDP/FiniteHorizonPOMDPs.jl/pull/8#issuecomment-775271381), but I am not sure if it's not causing problems elsewhere.
> 
> I strongly think we should NOT have the three argument `stage_stateindex(mdp, s, stage)`. The stage information should be included in the state.

@zsunberg I believe this corresponds to the structure suggested in former Readme.md file and the method itself was implemented by you. While I agree this is somewhat redundant information, I think this supports the idea that the method returns the local index for the given stage, thus supports readability.

zsunberg replied:
> I believe this corresponds to the structure suggested in former Readme.md file

It could be :smile: If that was there, I have changed my mind.

> While I agree this is somewhat redundant information, I think this supports the idea that the method returns the local index for the given stage, thus supports readability.

In my opinion he difference in name between `stateindex` and `stage_stateindex` makes this clear.

I think having the redundant information could lead to confusion. For example, what should happen if
```julia
stage_stateindex(m, s, 2)
```
if `stage(s)` is `3`. I.e. what should happen if the third argument disagrees with the second?

zsunberg replied:
(I am open to arguments for doing it the other way though)

Omastto1 replied:
I think that if the user is able to do `stage_stateindex(m, s, 2)` then he could be also able to define all states to a specific stage, say 2. So it depends on the point of view.
That said I understand your point and will change this.

BoZenKhaa replied:
I am in favor of 2-argument `stage_stateindex(m, s)` calling `stage(s)` internally as well.

I was trying to make a different point, I will try to better explain what I think might be an issue here.
My point here is that `policy` is defined as a matrix:
```
mutable struct FiniteHorizonValuePolicy{Q<:AbstractArray, U<:AbstractMatrix, P<:AbstractMatrix, A, M<:MDP} <: Policy
    qmat::Q
    util::U 
    policy::P 
    action_map::Vector{A}
    include_Q::Bool 
    m::M
end
```

with rows holding the policy for each stage:
```
function addstagerecord(fhpolicy::FiniteHorizonValuePolicy, qmat, util, policy, stage)    
    fhpolicy.qmat[stage, :, :] = qmat
    fhpolicy.util[stage, :] = util
    fhpolicy.policy[stage, :] = policy
    return fhpolicy
end
```

However, the `action` method accesses policy with scalar index (stateindex returns Integer, not a CartesianIndex):
```
function action(policy::FiniteHorizonValuePolicy, s::S) where S
    sidx = stateindex(policy.m, s)
    aidx = policy.policy'[sidx]
    return policy.action_map[aidx]
end
```

Indexing with scalalar into matrices works and I think it would work fine in MDPs where `stateindex` transforms to `stage_stateindex` easily with mod n_stages, but it will fail if the user-defined `stateindex` orders states differently. 


I see now that my recommendation was wrong, but I still don't think that the current code works correctly. I think that the new code for `action` should be
```
function action(policy::FiniteHorizonValuePolicy, s::S) where S
    sidx = stage_stateindex(policy.m, s)
    aidx = policy.policy[stage(s), sidx]
    return policy.action_map[aidx]
end
```

I think the reason the current tests don't catch this is either because the stateindex->stage_stateindex wraps nicely with mod or because in 1-d gridworld, the optimal policy doesn't contain a whole lot of different actions so the mismatched indexes still return the correct action.

PS: wouldn't it be better to store the policy in columns?

BoZenKhaa replied:
Also, I think it would be nice to have a small test MDP instance with different number of states in each stage (and different transition probs as well) to catch weird edge cases.

Omastto1 replied:
In my opinion, the `[stage(), stage_stateindex]` should work the same as `[stateindex]`.

Column-based store: I have to yet study this feature of Julia, but this is to be done.
A different number of states in each stage: This feature is desirable. However, it is not possible right now. I am creating an issue for that.

Given that each stage would have a different number of states, we would have to also update `FHPolicy`, because it initializes variables as matrices. With the different number of states, we would have to initialize them as `(horizon(m) + 1) x max_states_in_stage` or as another structure, presumably as some sort of dictionary.

BoZenKhaa replied:
> In my opinion, the [stage(), stage_stateindex] should work the same as [stateindex].

I think this works only when `stateindex(s) = (stage(s)-1)*length(stage_states) + stage_stateindex`, i.e. when the stateindex wraps "nicely" at the end of the rows. I assume that is why transpose of the matrix before indexing is needed, the scalar indexing works by columns, and the code wouldn't work with `'` removed.

`stateindex` is up to user to define. So for example, consider MDP with 3 states `s, s', s''`  in stage 1. In stage one, they had `state_stageindex` `1,2,3,`.  However, their `stateindex` could be completeley independent of the `state_stageindex`, for example `4, 1, 13`. Indexing with `stateindex` would then fetch actions for states in different stages. 

Let me know if I misunderstood how the indexing works here, it really seems to me that while the current code works with the existing test cases, it should be modified to accommodate less specialized user definitions of `stateindex`. I am not sure wheter `stateindex()` is used elsewhere, but it should be used carefully as user implementations may order states differently than `state_stageindex()`.

Otherwise I think the code looks great!

Omastto1 replied:
You are right.
I already changed the code, as this is more general.

Omastto1 committed:
add constructor for FHPolicy and use of horizon(m)

Omastto1 committed:
replaced m.horizon with horizon(m)

Omastto1 committed:
changed stage_stateindex definition

Omastto1 committed:
Docstring

Omastto1 committed:
action access changed

Omastto1 committed:
Added verbose check

Omastto1 committed:
Support for sparse states and column-wise update

Omastto1 committed:
Updates states count for dense states

Omastto1 commented:
Changed matrices operations to column-based and added possibility to make policy and utility matrices sparse.

zsunberg commented:
FYI, I don't think it makes a lot of sense to make these things sparse. I don't think it is very common to have a sparse value function, and a sparse policy doesn't make very much sense to have. I believe sparse matrices have longer look-up times.

zsunberg commented:
I think it is better to use dense matrices with `Union{Float64, Nothing}` and `Union{A, Nothing}` elements.

BoZenKhaa commented:
What is the benefit of having the type `Union{Float64, Nothing}`? Is it to highlight that some elements can have no value or does it help save memory? 

I can't come up with any use-case where sparse value function or policy could be helpful. I was thinking of an MDP with many dead end states at every stage, but I couldn't think of domain, so I agree that sparse doesn't actually make sense here.

On the other hand, having different number of states at each stage should be relatively common (e.g. policy for MDP with known initial state in stage 1). In that case, having the matrix dense could significantly increase the memory footprint. So maybe an array of arrays of different size of type `Union{Float64, Nothing}` would work better for these instances and shouldn't hinder other solutions too much?  I am thinking that memory might be more critical than speed for FH since each value in the value matrix is set only once and read at most as many times as there are states in previous stage, so the write and look-up speeds shouldn't be as important as they are in VI. 

However, I think that we would need some larger benchmark instances to test this. It might be better to merge now with dense `Union{Float64, Nothing}` and leave this for later.

Omastto1 commented:
My opinion on this is reverting changes connected with sparse matrices and merge.

zsunberg commented:
> What is the benefit of having the type Union{Float64, Nothing}? Is it to highlight that some elements can have no value or does it help save memory?

No, I don't think there is a memory benefit - it just forces errors if you access memory that you're not supposed to.

> So maybe an array of arrays of different size ... would work better for these instances and shouldn't hinder other solutions too much?

Yeah, I think you are right. I think a vector of vectors is best, and it doesn't need to have `Union{Float64, Nothing}` - I think it can just be `Float64`

zsunberg commented:
(but it would be OK to merge with dense matrices)

Omastto1 committed:
Sparse matrices revert

Omastto1 committed:
small fix

Omastto1 commented:
@zsunberg I removed updates connected with sparse structures and believe this to be ready to merge.

zsunberg commented:
Ok, I'm merging this since this is definitely the way forward. We can move to issues to discuss future changes on a more granular level.
